### DISCUSSION TITLE
tutorials/shaders/compute_shaders

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#17] [User-contributed notes]
# tutorials/shaders/compute_shaders

This tutorial will walk you through the process of creating a minimal compute shader. But first, a bit of background on compute shaders and how they work with Godot. A compute shader is a special t...

https://docs.godotengine.org/en/latest/tutorials/shaders/compute_shaders.html

**Comment by bitmammoth** [comment]
For c# you may need to specify the GodotCollections.Array namespace type in order to avoid the non generic system.collections species.

**Comment by Aman-Anas** [comment]
Hopefully this information will be helpful to people using compute shaders in C# with large arrays. I couldn't find much clear documentation online this, especially for C# in Godot.

While you still can't (at the time of this writing) call `BufferGetData()` into a pre-existing allocated array (it allocates internally instead), there are a few useful methods in C# for avoiding memory allocations + copying in other parts of the compute shader workflow. This can be useful for improving performance and memory usage.

Instead of using `Buffer.BlockCopy()` to convert in-between a byte array and a float array (or your own data structure), consider using `Span`s and `MemoryMarshal`. Here are some examples. 

> Be very careful with this power! Make sure your memory is laid out correctly. BlockCopy is pretty fast for most use cases, only use this if needed, e.g. you have large arrays.

---

From the "Provide Input Data" step, we can instead re-interpret the float array.
```cs
var input = new float[] { 1, 2, 3, 4, 5, 6, 7, 8, 9, 10 };
var inputBytes = MemoryMarshal.AsBytes(input.AsSpan());
```
Alternatively, store the data as a byte field and cast it before adding your data (good for some use cases)
```cs
var elements = 5;
var inputBytes = new byte[elements * sizeof(float)];
var input = MemoryMarshal.Cast<byte, float>(inputBytes);
input[0] = 1.0f; // Now inputBytes has its data changed!
```
---

Another useful trick is to use `MemoryMarshal` with structs. This can help clean up GLSL code and improve the workflow.
```cs
using System.Runtime.InteropServices;

// Pack= is how to do the alignment. 1 is the tightest,
// but make sure to check GLSL rules to make it match
// depending on your use case.
[StructLayout(LayoutKind.Sequential, Pack = 1)]
public struct CoolData
{
    //  public fields may be bad practice, but using for simplicity
    public int wheels; 
    public float position;
    public float hoursOfSleep;
}
// as byte array (single struct)
byte[] coolDataBytes = new byte[Marshal.SizeOf<CoolData>()];
ref var myData = ref MemoryMarshal.AsRef<CoolData>(coolDataBytes);
myData.wheels = 4;
// You can also use MemoryMarshal.Cast() if you had many structs in an array.
// (For example, many structs generated on the GPU and returned with rd.BufferGetData())
```

**Reply by BrassicaDev** [reply]
Hi, may i ask how to pass a 3D texture as an sampler3D in compute shader?
I can obtain Array<Image> object through FastNoiselite.GetImage3D(16, 16, 16), but i have no idea on how to pass it to the compute shader.

**Reply by Aman-Anas** [reply]
@BrassicaDev You should take a look at Sebastian Lague's marching cubes compute shader example. Usually, if you have a 3D array, you'll want to compress it down to a 1D array and use the height, width, length to index it. That can be passed to a compute shader the same as any other regular array.

**Reply by BrassicaDev** [reply]
I ‘ll try this, thank you

**Reply by hoootis** [reply]
did you get that working? if so, how?

**Reply by BrassicaDev** [reply]
> did you get that working? if so, how?

@hoootis Yes, and compress 3D array to 1D need some side information to let compute shader know what the original shape is, then i wrote a reIndex function to find the data through globalInvocationID. it's works pretty good.

**Reply by Clocktown** [reply]
@BrassicaDev I don't know why you got feedback like this. There shouldn't be a need to compress the 3D array to 1D. 3D textures and images are a thing. You'd specify a `depth` in the `RDTextureFormat` and set its `texture_type` to `RenderingDevice.TEXTURE_TYPE_3D`.

**Comment by bkorkmaz** [comment]
_A very simplified explanation of the shader :)_

```glsl
#[compute]
#version 450

// Compute Shader can process data in three dimensions. 
// This is essentially an efficient design intended for 1D, 2D, and 3D data structures.
// In our example, since we are processing a 1D array, we are only using the X axis.
//////////////////////////////////////////////////////////////////////////////////////////////// 
// Defining the axes we'll use on the GPU. Here, only the X axis. 2 threads on the X axis.
layout(local_size_x = 2, local_size_y = 1, local_size_z = 1) in;

// Connecting our data to the GPU.
// set = Data type
// binding = GPU binding point
// std430 = GPU memory structure standard.
// restrict buffer MyDataBuffer = Buffer with restricted access for optimization purposes.
layout(set = 0, binding = 0, std430) restrict buffer MyDataBuffer {
    float data[];
}
my_data_buffer;

void main() {
    //Multiply each element of the array by 2 and write the result back to the same element's position.
    my_data_buffer.data[gl_GlobalInvocationID.x] *= 2.0;
}
```

**Comment by codywohlers** [comment]
### Should we add a section about running it again, or continuously in each frame?

For someone new to Godot and compute shaders this is great, except the part where you want to run it every frame.

I learned from a youtube video ([5CKvGYqagyI](https://www.youtube.com/watch?v=5CKvGYqagyI)) that you only need to make the compute list again each time.  (Unless I'm wrong or there's more to consider).  I modified the above  tutorial by making these variables global to the Node:
```gdscript
## Local rendering device
var rd :RenderingDevice

## Shader storage buffer
var buffer :RID
var input_bytes :PackedByteArray
var output_bytes :PackedByteArray

# Compute Shader
var pipeline :RID 
var uniform_set :RID
```

and then moved the compute list creation to it's own function and called it at the start, and then every frame (with a `bool` status variable to not call it multiple times)
```gdscript
var shader_on :bool = false

func create_compute_list():
	var compute_list :int = rd.compute_list_begin()
	rd.compute_list_bind_compute_pipeline(compute_list, pipeline)
	rd.compute_list_bind_uniform_set(compute_list, uniform_set, 0)
	rd.compute_list_dispatch(compute_list, 5, 1, 1)
	rd.compute_list_end()

func _process(_delta: float):
	if not shader_on:
		shader_on = true
		rd.submit()
		rd.sync()

		# Read back the data from the buffer
		output_bytes = rd.buffer_get_data(buffer)

		# Do game logic with output_bytes here.
	
		input_bytes = output_bytes
		create_compute_list()
		shader_on = false
```

and in `compute_example.glsl` i only multiply by `1.0001` instead of `2.0` (so it doesn't instantly explode to `inf`).

**Reply by bit-wise** [reply]
Very cool, but were there no ill effects in skipping this step?
```
# Create a uniform to assign the buffer to the rendering device
var uniform := RDUniform.new()
uniform.uniform_type = RenderingDevice.UNIFORM_TYPE_STORAGE_BUFFER
uniform.binding = 0 # this needs to match the "binding" in our shader file
uniform.add_id(buffer)
var uniform_set := rd.uniform_set_create([uniform], shader, 0) # the last parameter (the 0) needs to match the "set" in our shader file
```
If not, great, if so, where do you think that code would go?

**Reply by Clocktown** [reply]
@bit-wise I assume he still has that code in `_ready` or somewhere. Once created, the uniform set can be continued to be used for as long as you want. Only if you want to change the uniforms do you need to recreate it. You also don't have to keep the RDUniforms alive because they are passed into the uniform set already. And for stuff that frequently changes, you would use Push Constants anyway.

**Comment by bit-wise** [comment]
I am new to compute shaders and Godot, and it wasn't very obvious from this tutorial where to put the GScript code. I'm guessing in func _ready()?

**Reply by natstott** [reply]
The creation of the computer shader and buffer goes in ready. (or in a function called from there)

If you want it to run every frame then everything after 
compute_list = rd.compute_list_begin()

needs to be in _process or some other script that is updated when you need it to.

**Comment by natstott** [comment]
The compute shader described above runs under a local rendering device. I just learnt that a compute shader can run in the global rendering device as part of the main render thread, using:

`rd = RenderingServer.get_rendering_device()`

in place of create_local_rendering_device() as above.

The quirk that blocked me for a while, until ClayJohn explained it, is that if you do this then you cant use submit() and sync() -  the compute shader will be run before or after the rendering pipeline as long as the compute list is created and dispatched.

This means it can access some render buffers if they are exposed. The only example I know is the buffer used by a multimesh which can be obtained by 

`meshbuffer=RenderingServer.multimesh_get_buffer_rd_rid(thismesh)`

which allows you to update the transforms of a multimesh using the compute shader.

**Reply by Clocktown** [reply]
@natstott good callout, I also found that most examples seem to go the local device route, even though for what I'd want my compute shaders to do, I would want to run it on the main device to share the resources, so I don't have to download and reupload them.

I actually went and threw together a little example for that type of usage here: https://github.com/Clocktown/Godot-ComputeShader-Image-TextureRect-Example

**Comment by HornsGuy** [comment]
Two things about the "Freeing Memory" (https://docs.godotengine.org/en/latest/tutorials/shaders/compute_shaders.html#freeing-memory) section:

1. There is no mention that the rendering device holds on to a lot of memory and must also be freed (I know this is true for C#, unsure if this is the case for GDScript)
2. Here would be the code snippets for freeing (plus or minus freeing the RenderingDevice, but it seems appropriate to include):

GDScript:
```
# Freeing RIDs to prevent memory leaks
rd.free_rid(pipeline)
rd.free_rid(pipeline)
rd.free_rid(pipeline)

# Rendering Device must also be freed
rd.free()
```

C#:
```
# Freeing RIDs to prevent memory leaks
rd.FreeRid(pipeline)
rd.FreeRid(pipeline)
rd.FreeRid(pipeline)

# Rendering Device must also be freed
rd.Free()
```

Note: Trying to free the uniform set in Godot 4.3 stable throws an error that the RID doesn't exist even though the doc says uniform sets must be freed. Therefore it has been excluded from the code snippets

**Reply by Clocktown** [reply]
@HornsGuy the current docs seem to clarify on why that happened in the past, apparently the Uniform Sets are already freed automatically if any RIDs associated with it (via Uniforms or the Shader) are freed. That explains why that error happens; so I guess we just don't have to free them as long as we free what they depend on. Alternatively, just free the uniform sets before you free shaders, textures, buffers, etc.

**Comment by foohines** [comment]
The Compute Shader HeightMap Demo has moved to [here](https://github.com/godotengine/godot-demo-projects/tree/master/compute/heightmap).

**Reply by Calinou** [reply]
Pull request opened:

- https://github.com/godotengine/godot-docs/pull/11174

**Comment by Pux333** [comment]
How to add #include to glsl compute shaders?

**Reply by Calinou** [reply]
This is already supported, but your included file must **not** end with `.glsl` (`.glsl.inc` is OK, as Godot will not attempt to import `.inc` files):

- https://github.com/godotengine/godot-proposals/issues/9592

**Comment by Nodragem** [comment]
I tried to follow this tutorial and when I get different issues depending on the rendering mode:
- in Forward+, I get an error when using `var output_bytes := rd.buffer_get_data(buffer)` that is `Buffer is either invalid or this type of buffer can't be retrieved.`
- in Mobile, I succeed to retrieve the data but then `rd.free_rid(uniform_set)` return an error, that is `servers/rendering/rendering_device.cpp:6422 - Attempted to free invalid ID: 677754429243392`

Note that, this tutorial is not really easy to follow if you are a beginner. As it does not tell you where to put your code.
So here is my code is that can help anyone:

```gdscript
@tool
extends Node

@export_tool_button("Run Comp. Shader", "PlayOverlay") var action1 = run_compute_shader
@export_tool_button("Retrieve Results", "Loop") var action2 = retrieve_compute_results

var rd:RenderingDevice
var buffer:RID
var pipeline:RID
var uniform_set:RID
var input:PackedFloat32Array

func run_compute_shader():
	print("Button clicked!")
	if rd != null:
		print("--> Rendering!")
		rd.submit()
		rd.sync()

func retrieve_compute_results():
	print("Button clicked!")
	if rd != null and buffer != null:
		# Read back the data from the buffer
		var output_bytes := rd.buffer_get_data(buffer)
		var output := output_bytes.to_float32_array()
		print("--> Input: ", input)
		print("--> Output: ", output)
		print("Freeing Buffer:")
		rd.free_rid(buffer)
		print("Freeing Pipeline:")
		rd.free_rid(pipeline)
		print("Freeing Uniform set:")
		rd.free_rid(uniform_set)

func _ready():
	rd = RenderingServer.create_local_rendering_device()
	# Load GLSL shader
	var shader_file := load("res://compute_example.glsl")
	var shader_spirv: RDShaderSPIRV = shader_file.get_spirv()
	var shader := rd.shader_create_from_spirv(shader_spirv)
	
	# Prepare our data. We use floats in the shader, so we need 32 bit.
	input = PackedFloat32Array([1, 2, 3, 4, 5, 6, 7, 8, 9, 10])
	var input_bytes := input.to_byte_array()
	
	# Create a storage buffer that can hold our float values.
	# Each float has 4 bytes (32 bit) so 10 x 4 = 40 bytes
	buffer = rd.storage_buffer_create(input_bytes.size(), input_bytes)
	
	# Create a uniform to assign the buffer to the rendering device
	var uniform := RDUniform.new()
	uniform.uniform_type = RenderingDevice.UNIFORM_TYPE_STORAGE_BUFFER
	uniform.binding = 0 # this needs to match the "binding" in our shader file
	uniform.add_id(buffer)
	uniform_set = rd.uniform_set_create([uniform], shader, 0) 
	# the last parameter (the 0) needs to match the "set" in our shader file
	
	# Create a compute pipeline
	pipeline = rd.compute_pipeline_create(shader)
	var compute_list := rd.compute_list_begin()
	rd.compute_list_bind_compute_pipeline(compute_list, pipeline)
	rd.compute_list_bind_uniform_set(compute_list, uniform_set, 0)
	rd.compute_list_dispatch(compute_list, 5, 1, 1)
	rd.compute_list_end()

```

**Reply by Nodragem** [reply]
It looks like I can' edit my comment :/ I just wanted to add that after I reversed back to Forward+, `rd.buffer_get_data(buffer)` is working again. I just have the `rd.free_rid(uniform_set)` not working now.

**Reply by Calinou** [reply]
> It looks like I can' edit my comment :/

To edit comments on Giscus, you need to open the GitHub Discussion page and use the 3 dots button on the top-right corner next to your comment.

To open the discussion page, click the "\<number\> comments" link at the top:

<img width="362" height="46" alt="image" src="https://github.com/user-attachments/assets/006e5d94-309b-4d42-aeab-7d7d6d7bd5ca" />

**Reply by maximilya** [reply]
First of all, thank you for providing a more complete code sample -- I too was somewhat confused about where the various pieces should go in order to get a minimal working example.

I believe the reason you're getting that error on attempting to free `uniform_set` is due to the following (from the documentation for `RenderingDevice.uniform_set_create` :

> This will be freed automatically when the `shader` or any of the RIDs in the `uniforms` is freed.

In particular, the value passed `uniforms` in our case is `[uniform]`, where the creation of `uniform` involved calling `uniform.add_id(buffer)` -- this means that if you call `rd.free_rid(buffer)` before `rd.free_rid(uniform_set)`, `uniform_set` is actually freed automatically per the above detail quoted from the documentation.

I appreciate that Godot is free and open-source, but I do wish these tutorials/examples included just a little bit more discussion of details like this (especially just a link to a _complete_ example for reference would be great), since they can be very difficult for beginners to track down.

**Comment by BuzzyBzzz** [comment]
If anyone is having trouble getting any of the Godot-specific directives (like `#[compute]`) to work with external IDEs, you can try adding a trailing comment to the line above your stage directive, like this:
```glsl
// Only works for RDShaderFile and your IDE must support backslashes \
#[compute]
#version 450
````
Godot will still compile fine as long as your shader file is a RDShaderFile.
If your IDE handles backslashes `\` then it shouldn't error either since the code should become:
```glsl
// #[compute]
#version 450
```

Extra Info & Use-cases:
Godot specific directives are more like import hints since they get consumed by RDShaderFile's parser when the shader is imported. Backslash resolution happens during GLSL preprocessing, before compilation and SPIR-V generation, which both happen after the RDShaderFile parser runs.

There are a couple other Godot specific hints that this works with.
I found these through the RDShaderFile source code. 
I'm not great at C++ so if there's anything I get wrong or missed just let me know so I can learn. 
(づ￣ 3￣)づ

`#include "res://path"` : Gets parsed by RDShaderFile first so it can be ignored by IDEs too.
```glsl
// IDE can't see this. \
#include "res://my_include.gdshaderinc"
```

`#[versions]` & `VERSION_DEFINES`: This is how you can create different defines for different reasons. Ex. Desktop vs Mobile
Normally when left unused a default version ("") is created for you with no defines set. Below is how you can use it to create a default version that instead has a custom define set.
In GLSL you'll declare your default version name (You can add more than 1), and their related defines.
```glsl
// Editor-only shader permutations + defines. Must be placed before any shader stages. \
#[versions]
// Default version \
default = "#define DEFAULT_VERSION";
// You can define more versions here.

/* Place a #[stage] & regular #version directive here. */

// This is how you declare your version defines. This line must be somewhere underneath a #[stage] directive (like #[compute].) \
VERSION_DEFINES

// You can use your DEFAULT_VERSION define like a normal define now. It'll only be defined in Godot so you could use it to branch #include paths or whatever else you need. EX.
#ifdef DEFAULT_VERSION
	#include "res://path.gdshaderinc"
#else
	#include "C:/absolute/path.gdshaderinc"
#endif
```
In GDScript when you get your spirv you'll want to put in a version name that matches the one you want to use, like this:
`var spirv: RDShaderSPIRV = shader_filee.get_spirv("default")`

As a side note, with the `#[version]` hint, I did notice that you can use `\n` to create additional defines inside of a singular version's define string:
```glsl
default = "#define DEFAULT\n#define ENABLED_LODS\n#define ETC";
```

In the parsed version of the script your version defines will be placed where VERSION_DEFINES is.
I see this as a way to get several version-specific defines, but since this is an insert, I imagine you could probably do lots of other stuff with it too.


None of this is `#[compute]` specific, but this page seems like the most visible place to share the info. 
GL coding! 👍

**Comment by Clocktown** [comment]
I put together an example for a different use-case: https://github.com/Clocktown/Godot-ComputeShader-Image-TextureRect-Example

It queues the compute shader every physics frame, shows how to use push constants and textures as images that can be written to. It also uses the main rendering device and the texture allocated by the RenderingDevice API is bound to a regular TextureRect node. This is what you'd want to do if a shader modifies some larger GPU resource that you want to access in normal rendering without downloading and reuploading.
